tool/tsh/tsh.go

```diff
@@ -754,44 +767,61 @@ func executeAccessRequest(cf *CLIConf) {
 	onStatus(cf)
 }
 
-func showNodes(nodes []services.Server, verbose bool) {
+func printNodes(nodes []services.Server, format string) error {
+	switch strings.ToLower(format) {
+	case teleport.Table:
+		printNodesAsTable(nodes, false)
+	case teleport.Verbose:
+		printNodesAsTable(nodes, true)
+	case teleport.JSON:
+		out, err := json.MarshalIndent(nodes, "", "  ")
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		fmt.Println(string(out))
+	case teleport.Names:
+		for _, n := range nodes {
+			fmt.Println(n.GetHostname())
+		}
+	default:
+		return trace.Errorf("unsupported format. try 'json', 'table', or 'names'")
+	}
+
+	return nil
+}
+
+func printNodesAsTable(nodes []services.Server, verbose bool) {
+	// Reusable function to get addr or tunnel for each node
+	getAddr := func(n services.Server) string {
+		if n.GetUseTunnel() {
+			return "⟵ Tunnel"
+		}
+		return n.GetAddr()
+	}
+
+	var t asciitable.Table
 	switch verbose {
 	// In verbose mode, print everything on a single line and include the Node
 	// ID (UUID). Useful for machines that need to parse the output of "tsh ls".
 	case true:
-		t := asciitable.MakeTable([]string{"Node Name", "Node ID", "Address", "Labels"})
+		t = asciitable.MakeTable([]string{"Node Name", "Node ID", "Address", "Labels"})
 		for _, n := range nodes {
-			addr := n.GetAddr()
-			if n.GetUseTunnel() {
-				addr = "⟵ Tunnel"
-			}
-
 			t.AddRow([]string{
-				n.GetHostname(), n.GetName(), addr, n.LabelsString(),
+				n.GetHostname(), n.GetName(), getAddr(n), n.LabelsString(),
 			})
 		}
-		fmt.Println(t.AsBuffer().String())
 	// In normal mode chunk the labels and print two per line and allow multiple
 	// lines per node.
 	case false:
-		t := asciitable.MakeTable([]string{"Node Name", "Address", "Labels"})
+		t = asciitable.MakeTable([]string{"Node Name", "Address", "Labels"})
 		for _, n := range nodes {
 			labelChunks := chunkLabels(n.GetAllLabels(), 2)
-			for i, v := range labelChunks {
-				var hostname string
-				var addr string
-				if i == 0 {
-					hostname = n.GetHostname()
-					addr = n.GetAddr()
-					if n.GetUseTunnel() {
-						addr = "⟵ Tunnel"
-					}
-				}
-				t.AddRow([]string{hostname, addr, strings.Join(v, ", ")})
-			}
+			t.AddRow([]string{n.GetHostname(), getAddr(n), strings.Join(labelChunks[0], ", ")})
+			t.AddRow([]string{"", "", strings.Join(labelChunks[1], ", ")})
```

Comment:
Looks this will panic on nodes with fewer than `3` labels and will omit labels on nodes with greater than `4` labels.

Reply:
I see now, I totally misunderstood what `chunkLabels` was doing, thanks